plugins/atomicfu/atomicfu-compiler/src/org.jetbrains.kotlinx.atomicfu.compiler/extensions/AtomicFUTransformerJsIr.kt

```diff
@@ -0,0 +1,473 @@
+            }
+        }
+        return super.visitBlock(block)
+    }
+
+    override fun visitReturn(returnExpr: IrReturn): IrExpression {
+        returnExpr.value = returnExpr.value.transformAtomicFunctionCall()
+        return super.visitReturn(returnExpr)
+    }
+
+    override fun visitTypeOperator(typeOp: IrTypeOperatorCall): IrExpression {
+        typeOp.argument = typeOp.argument.transformAtomicFunctionCall()
+        return super.visitTypeOperator(typeOp)
+    }
+
+    override fun visitSetVariable(setVar: IrSetVariable): IrExpression {
+        setVar.value = setVar.value.transformAtomicFunctionCall()
+        return super.visitSetVariable(setVar)
+    }
+
+    override fun visitSetField(setField: IrSetField): IrExpression {
+        setField.value = setField.value.transformAtomicFunctionCall()
+        return super.visitSetField(setField)
+    }
+
+    override fun visitVariable(declaration: IrVariable): IrStatement {
+        declaration.initializer = declaration.initializer?.transformAtomicFunctionCall()
+        return super.visitVariable(declaration)
+    }
+
+    override fun visitBranch(branch: IrBranch): IrBranch {
+        branch.apply {
+            condition = condition.transformAtomicFunctionCall()
+            result = result.transformAtomicFunctionCall()
+        }
+        return super.visitBranch(branch)
+    }
+
+    override fun visitElseBranch(branch: IrElseBranch): IrElseBranch {
+        branch.apply {
+            condition = condition.transformAtomicFunctionCall()
+            result = result.transformAtomicFunctionCall()
+        }
+        return super.visitElseBranch(branch)
+    }
+
+    private fun IrExpression.transformAtomicValueInitializer() =
+        when {
+            type.isAtomicValueType() -> getPureTypeValue()
+            type.isAtomicArrayType() -> buildPureTypeArrayConstructor()
+            type.isReentrantLockType() -> buildConstNull()
+            else -> this
+        }
+
+    private fun IrDeclaration.transformAtomicInlineDeclaration() {
+        if (this is IrFunction &&
+            isInline &&
+            extensionReceiverParameter != null &&
+            extensionReceiverParameter!!.type.isAtomicValueType()
+        ) {
+            val type = extensionReceiverParameter!!.type
+            val valueType = type.atomicToValueType()
+            val getterType = buildFunctionSimpleType(listOf(irBuiltIns.unitType, valueType))
+            val setterType = buildFunctionSimpleType(listOf(valueType, irBuiltIns.unitType))
+            val valueParametersCount = valueParameters.size
+            val extendedValueParameters = mutableListOf<IrValueParameter>().apply {
+                addAll(valueParameters)
+                add(buildValueParameter(GETTER, valueParametersCount, getterType))
+                add(buildValueParameter(SETTER, valueParametersCount + 1, setterType))
+            }
+            this as IrSimpleFunction
+            (descriptor as FunctionDescriptorImpl).initialize(
```

Comment:
Please avoid using of `DeclarationDescriptor` API

Reply:
Could you please tell me, how can I avoid using DeclarationDescriptor API here? 

The problem is that if I just change the parameters of the owner `IrFunction`  (`extensionReceiver` and `valueParameters`) then `ManglerChecker` fails at the comparison of the actual `Ir` signature and the one declared in the `Descriptor`.

Reply:
What kind of failure do you see?

Reply:
I fixed this by creating new transformed declarations with `WrappedSimpleFunctionDescriptor` (as I can not just set a new descriptor).